source/h1_connection.c

```diff
@@ -743,68 +688,56 @@ static struct aws_h1_stream *s_update_outgoing_stream_ptr(struct h1_connection *
         current = NULL;
     }
 
-    /* If current stream is NULL, do one of:
-     * - Look for next item in stream_list.
-     * - Client side: look in synced_data.pending_stream_list for more work
-     * - Server side: look in thread_data.waiting_stream_list for more work */
+    /* If current stream is NULL, look for more work:
+     * If next_node in stream_list exists, work on that.
+     * Else, look for more work in other lists.
+     *      If server: look in thread_data.waiting_stream_list.
+     *      If client: look in synced_data.pending_stream_list. */
     if (!current && !connection->thread_data.is_writing_stopped) {
         if (next_node && next_node != aws_linked_list_end(&connection->thread_data.stream_list)) {
             current = AWS_CONTAINER_OF(next_node, struct aws_h1_stream, node);
-
-        } else if (connection->base.server_data) {
-            /* server side should check the stream already has response or not
-             * Require a lock to prevent the user makes any change to the stream state */
+        } else /* Look for more work in other lists */ {
             /* BEGIN CRITICAL SECTION */
             s_h1_connection_lock_synced_data(connection);
-            while (!aws_linked_list_empty(&connection->thread_data.waiting_stream_list)) {
-                /* The front of waiting_stream_list is not ready to be sent */
-                if (!AWS_CONTAINER_OF(
-                         aws_linked_list_front(&connection->thread_data.waiting_stream_list),
-                         struct aws_h1_stream,
-                         node)
-                         ->synced_data.has_outgoing_response) {
-                    break;
+
+            if (connection->base.server_data) {
```

Comment:
It is weird, and me and Denke considered alternatives when we built it, and they were even more confusing.

So `thread_data.waiting_stream_list` is where streams live that don't have responses yet. That list is only modified from the tread.

The user does touch these streams from outside the thread when they create a response. But they don't touch the list that the stream is in, they only touch `synced_data.has_outgoing_response`, and queue the outgoing stream task to actually move the stream from one list to another on the channel-thread.

Reply:
Could you document this in the code a bit, so the next person doesn't have to drink their feelings away before proceeding? :)

Reply:
I just did a major simplification, because I doubted anyone but me and Denke maintaining this.

`thread_data.waiting_stream_list`: REMOVED. Server's only need to worry about the 1 `thread_data.stream_list`

`synced_data.pending_stream_list` RENAMED `synced_data.new_client_stream_list` to be clear that it's only used by clients, and only when the stream is brand new.

This vastly simplifies the logic by which we choose the next stream to work on.

The downside is that, a connection with more than stream at a time (pipelining) will do an O(N) scan over a list to find the next stream to work on. But my understanding is no one does pipelining in HTTP/1, so that makes it real-world O(1) 😎

thoughts?